[AMDGPU] Change default AMDHSA Code Object version to 5

saiislam opened the pull request:
Also update LIT tests and docs.
For more details, see
https://llvm.org/docs/AMDGPUUsage.html#code-object-v5-metadata

saiislam commented on clang/include/clang/Basic/TargetOptions.h:
```diff
@@ -88,7 +88,7 @@ class TargetOptions {
   };
   /// \brief Code object version for AMDGPU.
-  CodeObjectVersionKind CodeObjectVersion = CodeObjectVersionKind::COV_None;
+  CodeObjectVersionKind CodeObjectVersion = CodeObjectVersionKind::COV_5;
```
Default was never `none`. It anyway gets overwritten with `MarshallingInfoEnum<TargetOpts<"CodeObjectVersion">,"COV_5">` which was earlier COV4.

This was defined as COV_None in the commit [[NFC] Fix uninitalized member variable use in ASTReader::ParseTargetOptions()](https://github.com/llvm/llvm-project/commit/6f2a865d2f6bc426a61939a0a1acfcb25d5c1a18).

I changed it to cov_5 because it is the new default and leaving it as cov_none might create confusion.

jhuber6 reviewed (approved):
Assuming all the tests pass, this LG and has been a long time coming. Thanks for working on this.

dwblaikie commented:
(patches like this should probably be broken up - test changes to the defaults in lld and llvm for instance don't depend on the change to the clang driver which is the only real semantic change in this patch, right? So probably only change the semantics of clang, and the tests that need to change there - then follow-up with separate commits that update other test coverage to be closer to the current defaults)

JonChesterfield commented on llvm/test/CodeGen/AMDGPU/GlobalISel/non-entry-alloca.ll:
```diff
@@ -75,8 +75,8 @@ bb.2:
 }
-; DEFAULTSIZE: .amdhsa_private_segment_fixed_size 4112
-; DEFAULTSIZE: ; ScratchSize: 4112
+; DEFAULTSIZE: .amdhsa_private_segment_fixed_size 16
```
This seems a bit suspect. It used to be about 4k and is now 16. Are we out by a factor of 1024 somewhere?

jhuber6 replied:
My understanding is that it's supposed to be `0` if the backend could not statically determine it, so it's unlikely to be due to that change in COV5. Maybe it's not matching the full line of 16384 or similar?

JonChesterfield commented:
This is a wild amount of code churn from a trivial change. 10k lines of almost all noise. Means the chances of us noticing breakage in a code review tool is pretty low.

How about as a first patch we pass `-code-object=v4` or whatever syntax to essentially all the tests, then rebase this, so that we can get something approximating "this is the functional change, with the codegen change visible in these tests"?

In general it seems likely that a lot of tests are checking things they don't actually care about, probably because they're frequently generated by the python thing. Maybe some of the noise can be removed by tweaking the test generator script to emit checks that are insensitive to ABI version?

saiislam committed:
[AMDGPU] Change default AMDHSA Code Object version to 5

Also update LIT tests and docs.
For more details, see
https://llvm.org/docs/AMDGPUUsage.html#code-object-v5-metadata

Note: This is the first commit in this series which only handles
default change in clang and corresponding tests.

saiislam committed:
fixup! [AMDGPU] Change default AMDHSA Code Object version to 5

This is second commit in this series. It only deals with cov5
value change, implictarg size change, and corresponding test in
MLIR.

saiislam committed:
fixup! [AMDGPU] Change default AMDHSA Code Object version to 5

This is the third commit in this series and it only deals with
LLD test changes wrt cov5.

saiislam committed:
fixup! [AMDGPU] Change default AMDHSA Code Object version to 5

This is the fourth commit in this series. It only deals with changes
in llvm-objdump tests wrt cov5 default.

saiislam committed:
fixup! [AMDGPU] Change default AMDHSA Code Object version to 5

This is the fifth commit in this series. It only deals with changes
in llvm source code, documentation, and GlobalIsel wrt cov5 default.

saiislam committed:
fixup! [AMDGPU] Change default AMDHSA Code Object version to 5

This is the seventh commit in this series. It only deals with
manual changes in llvm lit tests wrt cov5 default.

saiislam committed:
fixup! [AMDGPU] Change default AMDHSA Code Object version to 5

This is the ninth and final commit in this series. It only deals
with autogenerated changes in llvm lit tests tests wrt cov5 default.

saiislam commented:
> (patches like this should probably be broken up - test changes to the defaults in lld and llvm for instance don't depend on the change to the clang driver which is the only real semantic change in this patch, right? So probably only change the semantics of clang, and the tests that need to change there - then follow-up with separate commits that update other test coverage to be closer to the current defaults)

Thanks for the suggestion! Based on @dwblaikie and @JonChesterfield comments, I have broken this PR into 7 commits. Each commit deals with a specific subset (clang, lld, mlir, llvm manual tests, llvm autogenerated tests, etc.).
Hopefully this will simplify reviewers' job.

I will squash and merge these 7 commits into a single commit at the time of merging.

yxsamliu reviewed (approved):
LGTM. Thanks

dwblaikie commented on clang/include/clang/Driver/Options.td:
```diff
@@ -4708,12 +4708,12 @@ defm amdgpu_ieee : BoolOption<"m", "amdgpu-ieee",
 
 def mcode_object_version_EQ : Joined<["-"], "mcode-object-version=">, Group<m_Group>,
-  HelpText<"Specify code object ABI version. Defaults to 4. (AMDGPU only)">,
+  HelpText<"Specify code object ABI version. Defaults to 5. (AMDGPU only)">,
```
Be nice if we could avoid having this written in 3 places... are there other techniques used for other flags with default values

dwblaikie commented on clang/lib/Driver/ToolChains/CommonArgs.cpp:
```diff
@@ -2402,7 +2402,7 @@ void tools::checkAMDGPUCodeObjectVersion(const Driver &D,
 unsigned tools::getAMDGPUCodeObjectVersion(const Driver &D,
                                            const llvm::opt::ArgList &Args) {
-  unsigned CodeObjVer = 4; // default
+  unsigned CodeObjVer = 5; // default
```
Perhaps at least this could be avoided? The flag configuration already says what th edefault value is - could that be relied on without needing to specify a value here too?

dwblaikie commented:
Generally when we split things up, they're separate reviews and separate commits. fixups are for necessary changes that need to be applied atomically (fixes to the base patch in a pull request, etc).

dwblaikie commented:
(part of the point is so that patches can be reverted as needed without having to revert/reapply huge patches when they aren't actually dependent)
